5_data_munge/src/munge_wqp_files.R

```diff
@@ -93,19 +93,123 @@ munge_wqp_withoutdepths <- function(in_ind, min_value, max_value, max_daily_rang
+           temperature_max_daily > min_value & temperature_max_daily < max_value|is.na(temperature_max_daily))
 
   # save
   data_file <- scipiper::as_data_file(out_ind)
   saveRDS(dat_daily, data_file)
   gd_put(out_ind)
 
 }
+
+resolve_statcodes <- function(in_ind, out_ind) {
+  dat <- readRDS(sc_retrieve(in_ind, remake_file = 'getters.yml')) %>%
+    ungroup() %>%
+  # drop values that are estimated or blank-corrected
+  # drop values that are not min, mean, max
+    filter(!ResultValueTypeName %in% c('Estimated', 'Blank Corrected Calc')) %>%
+    filter(StatisticalBaseCode %in% c(NA, 'Mean', 'Minimum', 'Maximum', 'mean',
+                                      'Geometric Mean', 'Daily Maximum', 'Daily Minimum',
+                                      'Daily Geometric Mean'))
+
+
+  # print message that says how many observations we lost when dropped
+  nrow_o <- nrow(readRDS(sc_retrieve(in_ind, remake_file = 'getters.yml')))
+  message(paste(nrow_o - nrow(dat), 'observations were dropped due to estimation, blank correction, or statcode that was not mean, min, max'))
+  # for some data, the start and end dates are different, and data providers
+  # seem to be using these as a date range of the whole dataset
+  # sometimes, the proper collection date is in the comment field
+  # we're dropping data that has a StatisticalBaseCode because we don't want
+  # values averaged over multiple days
+  range_dates <- filter(dat, !is.na(ActivityEndDate)) %>%
+    filter(ActivityStartDate != ActivityEndDate) %>%
+    filter(is.na(StatisticalBaseCode)) %>%
+    filter(grepl('Collected', ResultCommentText)) %>%
+    mutate(newActivityStartDate = gsub('(Collected: )(.*\\d{4})(\\s*\\d+.*)', '\\2', ResultCommentText, perl = TRUE),
+           newActivityStartTime = gsub('(Collected: .*\\d{4}\\s*)(\\d+.*)', '\\2', ResultCommentText, perl = TRUE)) %>%
+    mutate(newActivityStartTime = format(strptime(newActivityStartTime, format = '%I:%M %p'), '%H:%M:%S'),
+           newActivityStartDate = as.Date(newActivityStartDate, format = "%b %d %Y")) %>%
+    select(-ActivityStartDate, -`ActivityStartTime/Time`) %>%
+    rename(ActivityStartDate = newActivityStartDate,
+           `ActivityStartTime/Time` = newActivityStartTime)
+
+  # print message about date recoveries
+  message(paste(nrow(range_dates), 'observations with mismatching start/end dates were recovered by extracting collection dates from comments'))
+
+  # those that don't have collected in the comments
+  other <- filter(dat, !is.na(ActivityEndDate)) %>%
+    filter(ActivityStartDate != ActivityEndDate) %>%
+    filter(is.na(StatisticalBaseCode)) %>%
+    filter(!grepl('Collected', ResultCommentText))
+
+  # look at the number of obs per day per site-date to weed out bad sites
+  # only keep sites that have one obs per day
+  # site dates with > 1440 obs (which is 1obs/minute) was most of the drops here,
+  # so we're confident we're dropping bad data
```

Comment:
So what do you suspect is going on here? Are these cases where `ActivityStartDate` is just the start date of the monitoring program, and we cannot recover information about the actual date these samples were taken?

Reply:
Yes, that for some reason, they're using start/end dates as the _project_ start/end dates, and observation date can't be recovered anywhere. We keep the records where start/end dates are different, there is no stat code, and there is a single observation because we assume these are discrete observations.

Reply:
Also note, this is mostly a WI DNR issue. The output from the `message` call on line 194 is:

`3931 site-dates and 14703936 raw observations were dropped because n>1 obs per site-date and start-end dates did not match. WIDNR was responsible for 3291 site-dates and 14500047 raw observations.`